### PR TITLE
Added Editor Protocol option using babel option and env var.

### PR DESCRIPTION
**Pull request by ShubhamVerma1811**
Hey Sid,
This plugin is super useful.
I have added a feature for letting user enter their preferred editor.
This PR works on the points of #1 and #6.

This can be done via a babel option called "editorURLProtocol" or from an .env var.

Babel option has higher priority than env approach.
Please review the code and let me know if any futher changes are required.

Thanks!

### CODE REVIEW

**Comment by siddharthkp** [comment]
This is the way I'd prefer it is to just accept the `editor` name: vscode | phpstorm | sublime, etc. and hide the detail about how they open files. For example, phpstorm has a different string: https://github.com/siddharthkp/babel-plugin-open-source/issues/6#issue-999132767

I'm guessing the .env file is so that different team members can set their own preferred editor?

**Comment by ShubhamVerma1811** [comment]
Sure Sid, I'll work on the editor point.
 
And yes, the .env approach is for personalized editors.

**Comment by ShubhamVerma1811** [comment]
Hey again @siddharthkp,
As per the points you mentioned, I have added the supports for following editors with their own protocols.
- Sublime (Tested on Windows)
- VScode and VSCode Insiders (Tested on Windows)
- Atom (Tested on Windows)
- PhpStorm (Not Tested)

Usage would be like:
```      ["babel-plugin-open-source", {
        "editor":"vscode-insiders", // sublime | atom | phpstorm
        "envPath":"./.env"
      } ]
```

And now .`env` has higher prority over "editor" option.

Please review the code and let me know of other changes.
Thanks!

File: packages/babel-plugin-open-source/script.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ if (typeof document !== 'undefined') {
     if (!event.altKey) return;
 
     event.preventDefault();
-    const { filename, start, editor } = JSON.parse(event.target.dataset.source)
-    window.open(editor + '://file/' + filename + ':' + start)
+    const { filename, start, url } = JSON.parse(event.target.dataset.source)
```
Comment (siddharthkp) [inline packages/babel-plugin-open-source/script.js:7]: we can remove filename + start from the payload and plugin completely if it's unused now?

File: packages/babel-plugin-open-source/babel.js [hunk 2]
```diff
@@ -33,26 +33,46 @@ module.exports = declare(api => {
...
+        // https://github.com/siddharthkp/babel-plugin-open-source/issues/6#issue-999132767
+        url = `phpstorm://open?file=${state.filename}&line=${location.start.line}`
+      } else if (editor === 'atom') {
+        // https://flight-manual.atom.io/hacking-atom/sections/handling-uris/#core-uris
+        url = `atom://core/open/file?filename=${state.filename}&line=${location.start.line}`
+      } else if (editor === 'vscode-insiders') {
+        url = `vscode-insiders://file/${state.filename}:${location.start.line}`
+      } else {
+        url = `vscode://file/${state.filename}:${location.start.line}`
+      }
```
Comment (siddharthkp) [inline packages/babel-plugin-open-source/babel.js:68]: this is very well done!

File: packages/babel-plugin-open-source/babel.js [hunk 3]
```diff
@@ -33,26 +33,46 @@ module.exports = declare(api => {
...
+      let url = null;
+      let editor = state.opts?.editor?.toLowerCase() || 'vscode';
 
       // the element was generated and doesn't have location information
       if (!location) return
 
       try {
         if(state.opts.envPath) {
-          editorURLProtocol = dotenv.parse(fs.readFileSync(filePath.resolve(process.cwd(),state.opts.envPath), 'utf8'))?.BABEL_OPEN_SOURCE_EDITOR;
+          editor = dotenv.parse(fs.readFileSync(filePath.resolve(process.cwd(),state.opts.envPath), 'utf8'))?.BABEL_OPEN_SOURCE_EDITOR;
```
Comment (siddharthkp) [inline packages/babel-plugin-open-source/babel.js:44]: I think we should move this below the line that checks for Fragment return (line 52)

File: packages/babel-plugin-open-source/babel.js [hunk 4]
```diff
@@ -33,26 +33,46 @@ module.exports = declare(api => {
...
         }
-      }  catch (error) {}
+      }  catch (error) {
+        console.error("ERROR in Babel Plugin Open Source", error)
```
Comment (siddharthkp) [inline packages/babel-plugin-open-source/babel.js:47]: Would be nice to know what went wrong 😅 And how to recover from it

File: packages/babel-plugin-open-source/babel.js [hunk 5]
```diff
@@ -30,18 +33,46 @@ module.exports = declare(api => {
...
 
       const location = path.container.openingElement.loc
+      let url = null;
+      let editor = state.opts?.editor?.toLowerCase() || 'vscode';
 
       // the element was generated and doesn't have location information
       if (!location) return
 
+      try {
+        if(state.opts.envPath) {
```
Comment (siddharthkp) [inline packages/babel-plugin-open-source/babel.js:43]: Feels a bit odd to ask for an `envPath`. Is it safe to follow the popular convention of finding a `.env` file at the root of the repository?

Comment (siddharthkp) [reply packages/babel-plugin-open-source/babel.js:43]: Sidenote: something to consider: Is `.editorconfig` the right place to add our custom key? (https://editorconfig.org)

File: packages/babel-plugin-open-source/babel.js (hunk 2 above)
Comment (ShubhamVerma1811) [reply packages/babel-plugin-open-source/babel.js:68]: Thanks a lot!

File: packages/babel-plugin-open-source/babel.js (hunk 5 above)
Comment (ShubhamVerma1811) [reply packages/babel-plugin-open-source/babel.js:43]: Yeah I can try the root .env approach.

For `.editorconfig`, it is not personalized right?
I mean it is pushed the GitHub so it's not unique..

Whereas .env aren't pushed, hence they can contain unique values..

Makes sense? 😅

File: packages/babel-plugin-open-source/babel.js (hunk 4 above)
Comment (ShubhamVerma1811) [reply packages/babel-plugin-open-source/babel.js:47]: The error would from path.resolve saying .env not found..

On error, it would anyways pickup vscode as default editor.

But I can surely right a helpful message with the error.

File: packages/babel-plugin-open-source/babel.js (hunk 5 above)
Comment (siddharthkp) [reply packages/babel-plugin-open-source/babel.js:43]: Makes sense!

didn't know .editorConfig are pushed to github 👍

Comment (ShubhamVerma1811) [reply packages/babel-plugin-open-source/babel.js:43]: Even I am not sure if they are supposed to be pushed.
I have just noticed them in some places on GitHub.

**Comment by ShubhamVerma1811** [comment]
Made the required changes.
`dotenv.config()` would pick the root folder's `.env`.
If editor exists in `.env` then that is used ,else from the `opts.editor` or `vscode`

**Comment by siddharthkp** [comment]
Last thing required here is an update to the readme :)

File: packages/babel-plugin-open-source/README.md [hunk 6]
```diff
@@ -23,11 +23,25 @@ Add the plugin to your babel config.
 // .babelrc
 {
    "plugins": [
-      "babel-plugin-open-source"
+      "babel-plugin-open-source",
+      "editor:"vscode", // vscode-insiders | sublime | atom
```
Comment (siddharthkp) [inline packages/babel-plugin-open-source/README.md:27]:
```suggestion
      "editor:"vscode", // optional, default: "vscode". options: vscode | vscode-insiders | sublime | atom | phpstorm
```

File: packages/babel-plugin-open-source/README.md [hunk 7]
```diff
@@ -23,11 +23,25 @@ Add the plugin to your babel config.
...
     ]
 }
 ```
 
+Current Supoorted editors are:
```
Comment (siddharthkp) [inline packages/babel-plugin-open-source/README.md]: I'd remove this list in favor of the exact string values in the suggestion above ⬆️

File: packages/babel-plugin-open-source/README.md [hunk 8]
```diff
@@ -23,11 +23,25 @@ Add the plugin to your babel config.
...
+- VSCode
+- VSCode-Insiders
+- Sublime
+- Atom
+- Phpstorm
+
+Personalized editors
```
Comment (siddharthkp) [inline packages/babel-plugin-open-source/README.md:39]: Can we rephrase this as - "If folks in your team use different editors, you can customise the editor just for you by adding ..."

Comment (ShubhamVerma1811) [reply packages/babel-plugin-open-source/README.md]: Pushed the changes.

**Review by siddharthkp — APPROVED** [review]
Let's ship it!

**Comment by ShubhamVerma1811** [comment]
Let's go! 🎉

**Comment by siddharthkp** [comment]
Thanks for the contribution! And that's for taking my annoyingly pin-perfect feedback 😅
